### DISCUSSION TITLE
SerialStringWrite error

### DISCUSSION THREAD

**Discussion by horsto** [bonsai-rx/bonsai#956] [Q&A] [answered]
I encounter errors when using `SerialStringWrite` with a microcontroller. This is a follow-up to https://github.com/bonsai-rx/bonsai/discussions/953 - I am now trying to write values of keys I pressed to the microcontroller. 
<img width="385" alt="workflow2" src="https://user-images.githubusercontent.com/7548928/177013983-1c3b0d1a-36b7-4ec3-b4f5-a15f3b20d293.PNG">

When using a simple `String` as input to `SerialStringWrite` everything works: The microcontroller receives the string and does what it is supposed to do. However, when using the workflow depicted above, I run into the error message `The given port name does not start with COM/com or does not resolve to a valid serial port. Parameter name: portName`. I am changing nothing else but the input to the `SerialStringWrite` node. 

This looks like an initialization problem to me. But I don't know how to solve it. Any ideas?

**Comment by horsto** [comment]
Basically, whenever I try to embed the SerialStringWrite into "something" in my main workflow, it throws an error upon start. It does not matter where I interface or when - only if I keep it separate from the main workflow it works

**Reply by horsto** [reply]
Another issue is that I want the serial write to happen only (and only when) a key has been pressed. It should not engage the serial connection in the meantime (i.e. no key pressed). Sorry for these naive (beginner) questions!

**Comment by bruno-f-cruz** [comment]
Could you include the workflow? thanks

**Reply by horsto** [reply]
See attached. This runs into the error described above. 
I want to broadcast the value attached to the keypress both into the running camera timestamp log as well as to the microcontroller. 
[keylogging.zip](https://github.com/bonsai-rx/bonsai/files/9035448/keylogging.zip)

**Comment by horsto** [comment]
It does actually work if I just turn the publishing sequence around: I.e. I have the key presses attached to the serial write procedure and just publish this into the video / frame logger. 
[keylogging.zip](https://github.com/bonsai-rx/bonsai/files/9035474/keylogging.zip)

**Reply by horsto** [reply]
Update: No, this does not work after restarting Bonsai... I am confused.

**Comment by horsto** [comment]
Not to add more confusion here, but I also tried to switch to OSC as an alternative messaging service - to have a python script listen to what comes out of Bonsai. The same "flavour" of error occurs also with the OSC nodes when I put them into my workflow. Then the error is "The specified connection has no matching configuration". So in each case the initialization fails because the port / ... cannot be found. What am I doing wrong? I am attaching the OSC example (just a drop in replacement for the serial connection) here too. 
[keylogging.zip](https://github.com/bonsai-rx/bonsai/files/9035602/keylogging.zip)

**Comment by bruno-f-cruz** [comment, marked as answer]
Could you give this modified version a try?
[keyloggingFix.zip](https://github.com/bonsai-rx/bonsai/files/9038289/keyloggingFix.zip)

**Reply by horsto** [reply]
This works. Magic to me since I don't understand the details. Thanks a lot for your help!

**Reply by bruno-f-cruz** [reply]
You gotta initialize the Serial object first. Otherwise funny things like this happen :P. Using subjects ensures that.

**Reply by horsto** [reply]
Thanks @bruno-f-cruz !
